### PR TITLE
Grocery Store Submission - Ren Carothers

### PR DESCRIPTION
**Pull request by RenCarothers**
# Assignment Submission: Grocery Store

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | This throws an error that stops the program and provides error type details.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | Class methods let us easily search through all the instances created of the class.  Instance methods would be suited for searching (or other action) within an instance.
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | A customer can have zero/many orders (one to many) and an order only has one customer (one to one).  One Solar System could have many planets (or zero I suppose), and a single planet belonged only to one system.
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | In the order CSV file, each order has a customer id.  In the customer csv, there is not order information.   In my program, I made each order instance contain as its customer details the customer instance, so that any time I wanted to call info about an order, among the things I could reference, was any of the customer details.  Whereas in the files, I would have had to switch between the two files to find all the customer info for an order.
Did the presence of automated tests change the way you thought about the problem? How? | The tests helped to to have a direction to write my code, to make sure I was accounting for what was being tested.  It provided helpful direction.  When I got to the sections of writing my own tests, ideally I would have written the tests and then created the code to pass, but I am not totally used to doing this yet and I sort of stumbled through it.  I more or less wrote the code, tested with puts as I went and afterwards, thought about edge cases and wrote tests to test for them. Eep.  Sorry I'll do better next time haha.  It's hard for me to visualize without puts as I go.

### CODE REVIEW

File: lib/customer.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+require 'csv'
+
+class Customer
+
+  def initialize(id, email, address)
+  @id = id
+  @email = email
+  @address = address
+  end
```
Comment (CheezItMan) [inline lib/customer.rb:9]: Indentation

```suggestion
  def initialize(id, email, address)
    @id = id
    @email = email
    @address = address
  end
```

File: lib/customer.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  attr_reader :id
+  attr_accessor :email, :address
```
Comment (CheezItMan) [inline lib/customer.rb:12]: These should be attr_reader instead of attr_accessor because you don't need to give users of your class the ability to directly change the instance variables.  A reader is more than enough access.

File: lib/customer.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  def self.all
```
Comment (CheezItMan) [inline lib/customer.rb:14]: 👍

File: lib/customer.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      customer_array << new_customer
+    end
+
+    return customer_array
+
+    # return array of instances
+    # to be invoked by self.find
+  end
+
+  def self.find(id_handed)
```
Comment (CheezItMan) [inline lib/customer.rb:38]: 👍

File: lib/customer.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        return customer_instance
+      end
+    end
+
+    # if id matches customer, return customer
+    # not parsing csv, parse arr!
+    # return customer
+  end
+
+  def self.save (filename, new_customer)
```
Comment (CheezItMan) [inline lib/customer.rb:52]: 👍 , nice the extra!

File: lib/order.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
+require 'csv'
+require_relative 'customer'
+
+class Order
+  attr_reader :id
+  attr_accessor :products, :customer, :fulfillment_status
```
Comment (CheezItMan) [inline lib/order.rb:6]: Similar to the issues above.

File: lib/order.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+  def initialize(id, products, customer, fulfillment_status = :pending)
+    @id = id
+    @products = products
+    @customer = customer
+    @fulfillment_status = fulfillment_status
+    raise ArgumentError.new("status must be pending, paid, processing, shipped or complete") unless ACCEPTABLE_FULFILLMENT_STATUS.include?(fulfillment_status)
+  end
+
+  def total
```
Comment (CheezItMan) [inline lib/order.rb:19]: 👍

File: lib/order.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    output = (cost_sum * 1.075).round(2)
+    return output
+  end
+
+  def add_product(name, price)
+    raise ArgumentError.new("Product already exists") if @products.has_key?(name)
+    @products[name] = price
+  end
+
+  def remove_product(name)
```
Comment (CheezItMan) [inline lib/order.rb:31]: 👍

File: lib/order.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  def total
+    cost_sum = 0
+    products.sum {|k,v| cost_sum += v}
+    output = (cost_sum * 1.075).round(2)
+    return output
+  end
+
+  def add_product(name, price)
```
Comment (CheezItMan) [inline lib/order.rb:26]: 👍

File: lib/order.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  end
+
+  def self.all
+    order_array = []
+
+    CSV.read('data/orders.csv').each do |row_entry|
+    products_hash = {}
+    row_entry[1].split(';') do |product|
+          products_hash[product.split(':')[0]] = product.split(':')[1].to_f
+    end
```
Comment (CheezItMan) [inline lib/order.rb:43]: nice work splitting up the product string. 

This might be handy to make a method, just to make the method simpler and easier to read.

File: lib/order.rb [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    new_order = Order.new(row_entry[0].to_i, products_hash, Customer.find((row_entry[2]).to_i), row_entry[3].to_sym)
+    order_array << new_order
+    end
+
+    return order_array
+  end
+
+  def self.find(id_handed) # instructions were to find AN instance so I did that, but in the optional find_by_customer, the method finds all
```
Comment (CheezItMan) [inline lib/order.rb:52]: 👍

File: lib/order.rb [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    self.all.find do |order_instance|
+      if order_instance.id == id_handed
+        return order_instance
+      end
+    end
+  end
+
+  def self.find_by_customer(customer_id)
```
Comment (CheezItMan) [inline lib/order.rb:60]: 👍

File: test/order_test.rb [hunk 13]
```diff
@@ -111,13 +111,53 @@
       expect(order.total).must_equal before_total
     end
   end
+
+  describe "#remove_product" do
```
Comment (CheezItMan) [inline test/order_test.rb:115]: 👍

**Review by CheezItMan — COMMENTED** [review]
# Grocery Store





## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The `Customer` and `Order` classes have appropriate `attr_` helpers for instance variables | ✔️, but default to making them attr_reader instead of attr_accessor.  Only make it an attr_accessor if you *need* to give users access to change the instance variables. 
The `CSV` library is used in `.all` methods, and not in `.find` methods | ✔️
Appropriately parses the product data from CSV file in `Order.all` | ✔️
`Order.all` calls `Customer.find` to set up the composition relation | ✔️
Practices git with at least 3 small commits and meaningful commit messages | ✔️, see my notes below on commit messages.  Wave numbers do not make good commit messages.

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
All of the wave 1 tests in `Customer` pass | ✔️
All of the wave 1 tests in `Order` pass | ✔️
All of the wave 2 tests in `Customer` pass | ✔️
All of the wave 2 tests in `Order` pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 4+ in Code Review && 3+ in Functional Requirements | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

Nice work, you hit the learning goals here.  I like the tests you wrote, they look like they work well.
